rust/automerge/src/op_tree.rs

```diff
@@ -71,78 +190,200 @@ impl OpTreeInternal {
+            }
+
+            if op.overwrites(e) {
+                succ.push(pos);
+
+                if e.visible() {
+                    overwritten = Some(e);
+                }
+            } else if e.visible() {
+                if found.is_none() && overwritten.is_none() {
+                    before = Some(e);
+                    num_before += 1;
+                } else {
+                    after = Some(e);
+                }
+            }
+
+            pos += 1;
+            next = iter.next();
+        }
+
+        pos = found.unwrap_or(pos);
+
+        SeekOpFound {
+            before,
+            num_before,
+            after,
+            overwritten,
+            succ,
+            pos,
+            index,
         }
     }
 
-    pub(crate) fn map_range<'a, R: RangeBounds<String>>(
+    pub(crate) fn seek_opid<'a>(
         &'a self,
-        range: R,
-        meta: &'a OpSetMetadata,
-    ) -> Option<query::MapRange<'a, R>> {
-        if self.root_node.is_some() {
-            Some(query::MapRange::new(range, self, meta))
-        } else {
-            None
+        opid: OpId,
+        encoding: ListEncoding,
+        clock: Option<&Clock>,
+        meta: &OpSetMetadata,
+    ) -> Option<(&'a Op, usize, bool)> {
+        let query = self.search(query::OpIdSearch::opid(opid, encoding, clock), meta);
+        let pos = query.found()?;
+        let mut iter = self.iter();
+        let op = iter.nth(pos)?;
+        let index = query.index_for(op);
+        for e in iter {
+            if e.elemid_or_key() != op.elemid_or_key() {
+                break;
+            }
+
+            if e.visible() {
+                return Some((op, index, false));
+            }
         }
+        Some((op, index, op.visible()))
     }
 
-    pub(crate) fn map_range_at<'a, R: RangeBounds<String>>(
+    pub(crate) fn seek_op_with_observer<'a>(
         &'a self,
-        range: R,
-        meta: &'a OpSetMetadata,
-        clock: Clock,
-    ) -> Option<query::MapRangeAt<'a, R>> {
-        if self.root_node.is_some() {
-            Some(query::MapRangeAt::new(range, self, meta, clock))
+        op: &'a Op,
+        encoding: ListEncoding,
+        meta: &OpSetMetadata,
+    ) -> SeekOpFound<'a> {
+        if let Key::Seq(_) = op.key {
+            let query = self.search(query::OpIdSearch::op(op, encoding), meta);
+            let pos = query.pos();
+            let index = query.index();
+            self.seek_update_for_op_for_observer(meta, op, pos, index)
         } else {
-            None
+            let pos = self.binary_search_by(|o| meta.key_cmp(&o.key, &op.key));
+            self.seek_update_for_op_for_observer(meta, op, pos, 0)
         }
     }
 
-    pub(crate) fn list_range<R: RangeBounds<usize>>(
-        &self,
-        range: R,
-    ) -> Option<query::ListRange<'_, R>> {
-        if self.root_node.is_some() {
-            Some(query::ListRange::new(range, self))
+    pub(crate) fn seek_op_simple<'a>(
```

Comment:
As per other comments I think this would be better named `find_op_without_observer`